File: tools/gerar_aceite_uat.py
# -*- coding: utf-8 -*-
"""Gera o Termo de Aceite de Testes (SIT/UAT) em .docx — documento de sign-off
que serve de gate para autorizar a virada oficial.

Uso:
    python gerar_aceite_uat.py [data/exemplo_cliente.yaml] [data/roteiros_teste.yaml]
"""
import os
import sys
from docx import Document
from docx.shared import Pt, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
import _common as C


def _shade_header(row):
    from docx.oxml.ns import qn
    from docx.oxml import OxmlElement
    for cell in row.cells:
        tcPr = cell._tc.get_or_add_tcPr()
        shd = OxmlElement("w:shd")
        shd.set(qn("w:val"), "clear")
        shd.set(qn("w:fill"), "1F4E78")
        tcPr.append(shd)
        for p in cell.paragraphs:
            for r in p.runs:
                r.font.color.rgb = RGBColor(0xFF, 0xFF, 0xFF)
                r.font.bold = True


def main(cliente_path="data/exemplo_cliente.yaml", roteiros_path="data/roteiros_teste.yaml"):
    cli = C.load_yaml(os.path.basename(cliente_path))
    rot = C.load_yaml(os.path.basename(roteiros_path))
    cliente = cli.get("cliente", {})
    projeto = cli.get("projeto", {})
    modulos = rot.get("modulos", [])

    doc = Document()

    h = doc.add_heading("TERMO DE ACEITE DE TESTES (SIT / UAT)", level=0)
    h.alignment = WD_ALIGN_PARAGRAPH.CENTER

    sub = doc.add_paragraph()
    sub.alignment = WD_ALIGN_PARAGRAPH.CENTER
    r = sub.add_run(f"{cliente.get('nome','')} — Projeto nº {projeto.get('numero','')}")
    r.italic = True

    doc.add_paragraph()
    doc.add_paragraph(
        "Este termo registra a conclusão e o aceite dos testes do sistema SIGER® para o cliente "
        f"{cliente.get('nome','')}, abrangendo o Teste Integrado (SIT) e o Aceite do Usuário (UAT) "
        "dos módulos no escopo. O aceite formaliza que os processos foram validados e autoriza a "
        "preparação da virada oficial."
    )

    # Dados
    doc.add_heading("Identificação", level=1)
    t = doc.add_table(rows=0, cols=2); t.style = "Table Grid"
    for k, v in [
        ("Cliente", cliente.get("nome", "")),
        ("Código SICLA", cliente.get("codigo_sicla", "")),
        ("RNS de Implantação", cliente.get("rns_implantacao", "")),
        ("Consultor responsável", projeto.get("consultor_responsavel", "")),
        ("Virada prevista", cliente.get("data_virada_prevista", "")),
        ("Data do aceite", C.today()),
    ]:
        cells = t.add_row().cells
        cells[0].paragraphs[0].add_run(k).bold = True
        cells[1].text = str(v)

    # Resumo por módulo
    doc.add_heading("Resultado por módulo", level=1)
    tm = doc.add_table(rows=1, cols=5); tm.style = "Table Grid"
    hdr = tm.rows[0].cells
    for i, txt in enumerate(["Módulo", "Casos", "Aprovados", "Reprovados", "Pendentes"]):
        hdr[i].paragraphs[0].add_run(txt)
    _shade_header(tm.rows[0])
    for mod in modulos:
        n = len(mod.get("casos", []))
        cells = tm.add_row().cells
        cells[0].text = mod.get("nome", "")
        cells[1].text = str(n)
        cells[2].text = ""; cells[3].text = ""; cells[4].text = ""
    doc.add_paragraph(
        "Preencher Aprovados/Reprovados/Pendentes a partir da planilha de roteiros (aba “Resumo e Sign-off”)."
    ).italic = True

    # Critério de liberação
    doc.add_heading("Critério de liberação (gate da virada)", level=1)
    for item in [
        "≥ 95% dos casos UAT com status Aprovado.",
        "Zero defeitos de severidade Crítica em aberto.",
        "Defeitos de severidade Alta com plano de ação acordado.",
        "Remessas bancárias e integrações com terceiros homologadas (quando aplicável).",
    ]:
        doc.add_paragraph(item, style="List Bullet")

    # Assinaturas
    doc.add_heading("Assinaturas", level=1)
    doc.add_paragraph()
    for papel in ["Consultor de Implantação — Rech", "Usuário Líder — " + cliente.get("nome", ""),
                  "Gerente do Projeto — Rech"]:
        doc.add_paragraph("__________________________________________")
        p = doc.add_paragraph(papel); p.runs[0].italic = True
        doc.add_paragraph()

    C.ensure_out()
    fname = f"Termo_Aceite_UAT_{C.slug(cliente.get('nome'))}.docx"
    path = os.path.join(C.OUT, fname)
    doc.save(path)
    print(f"OK: {fname} -> {path}")


if __name__ == "__main__":
    args = sys.argv[1:]
    main(*args) if args else main()

File: backend/src/geradores/gerar-aceite-uat.ts
import {
  AlignmentType,
  Document,
  HeadingLevel,
  Packer,
  Paragraph,
  Table,
  TableRow,
  TextRun,
  WidthType,
} from 'docx';
import { join } from 'path';
import { mkdirSync, writeFileSync } from 'fs';
import { OUT_DIR, carregarYaml, hoje, slug } from './comum';
import { GRADE, celula, celulaRotulo, tabelaComCabecalho } from './comum-docx';

/** Porte de `tools/gerar_aceite_uat.py` para Node/TS (§4.2/§4.7 dos Padrões da Rech).
 *
 * Gera o Termo de Aceite de Testes (SIT/UAT) em .docx — o documento de sign-off que serve de
 * gate para autorizar a virada oficial. Porte de EQUIVALÊNCIA — prova em
 * `gerar-aceite-uat.spec.ts`. */

interface Cliente {
  nome?: string;
  codigo_sicla?: string;
  rns_implantacao?: string;
  data_virada_prevista?: string;
}
interface Projeto {
  numero?: string | number;
  consultor_responsavel?: string;
}
interface DadosCliente {
  cliente?: Cliente;
  projeto?: Projeto;
}
interface Modulo {
  nome?: string;
  casos?: unknown[];
}
interface DadosRoteiros {
  modulos?: Modulo[];
}

/** Monta o documento. Separado de `gerar()` para o teste inspecionar sem gravar. */
export function montarDocumento(
  cli: DadosCliente = carregarYaml<DadosCliente>('exemplo_cliente.yaml'),
  rot: DadosRoteiros = carregarYaml<DadosRoteiros>('roteiros_teste.yaml'),
): Document {
  const cliente = cli.cliente ?? {};
  const projeto = cli.projeto ?? {};
  const modulos = rot.modulos ?? [];
  const nome = cliente.nome ?? '';

  const identificacao = new Table({
    width: { size: 100, type: WidthType.PERCENTAGE },
    borders: GRADE,
    rows: (
      [
        ['Cliente', nome],
        ['Código SICLA', cliente.codigo_sicla ?? ''],
        ['RNS de Implantação', cliente.rns_implantacao ?? ''],
        ['Consultor responsável', projeto.consultor_responsavel ?? ''],
        ['Virada prevista', cliente.data_virada_prevista ?? ''],
        ['Data do aceite', hoje()],
      ] as [string, string][]
    ).map(
      ([k, v]) =>
        new TableRow({ children: [celulaRotulo(k), celula(String(v))] }),
    ),
  });

  // Aprovados/Reprovados/Pendentes saem em branco de propósito: são preenchidos à mão a
  // partir da aba "Resumo e Sign-off" da planilha de roteiros.
  const porModulo = tabelaComCabecalho(
    ['Módulo', 'Casos', 'Aprovados', 'Reprovados', 'Pendentes'],
    modulos.map((mod) => [
      mod.nome ?? '',
      String((mod.casos ?? []).length),
      '',
      '',
      '',
    ]),
  );

  const criterios = [
    '≥ 95% dos casos UAT com status Aprovado.',
    'Zero defeitos de severidade Crítica em aberto.',
    'Defeitos de severidade Alta com plano de ação acordado.',
    'Remessas bancárias e integrações com terceiros homologadas (quando aplicável).',
  ];

  const papeis = [
    'Consultor de Implantação — Rech',
    `Usuário Líder — ${nome}`,
    'Gerente do Projeto — Rech',
  ];

  return new Document({
    sections: [
      {
        children: [
          new Paragraph({
            text: 'TERMO DE ACEITE DE TESTES (SIT / UAT)',
            heading: HeadingLevel.TITLE,
            alignment: AlignmentType.CENTER,
          }),
          new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [
              new TextRun({
                text: `${nome} — Projeto nº ${projeto.numero ?? ''}`,
                italics: true,
              }),
            ],
          }),
          new Paragraph({}),
          new Paragraph(
            'Este termo registra a conclusão e o aceite dos testes do sistema SIGER® para o cliente ' +
              `${nome}, abrangendo o Teste Integrado (SIT) e o Aceite do Usuário (UAT) ` +
              'dos módulos no escopo. O aceite formaliza que os processos foram validados e autoriza a ' +
              'preparação da virada oficial.',
          ),
          new Paragraph({
            text: 'Identificação',
            heading: HeadingLevel.HEADING_1,
          }),
          identificacao,
          new Paragraph({
            text: 'Resultado por módulo',
            heading: HeadingLevel.HEADING_1,
          }),
          porModulo,
          // O original faz `.italic = True` no parágrafo devolvido por add_paragraph(); em
          // python-docx isso só cria um atributo solto, sem efeito no documento. Como o porte
          // é de equivalência, reproduzimos o comportamento REAL (sem itálico), não a intenção.
          new Paragraph(
            'Preencher Aprovados/Reprovados/Pendentes a partir da planilha de roteiros (aba “Resumo e Sign-off”).',
          ),
          new Paragraph({
            text: 'Critério de liberação (gate da virada)',
            heading: HeadingLevel.HEADING_1,
          }),
          ...criterios.map(
            (t) => new Paragraph({ text: t, bullet: { level: 0 } }),
          ),
          new Paragraph({
            text: 'Assinaturas',
            heading: HeadingLevel.HEADING_1,
          }),
          new Paragraph({}),
          ...papeis.flatMap((papel) => [
            new Paragraph('__________________________________________'),
            new Paragraph({
              children: [new TextRun({ text: papel, italics: true })],
            }),
            new Paragraph({}),
          ]),
        ],
      },
    ],
  });
}

/** Gera o arquivo em `exemplos/`, com o mesmo nome do original. */
export async function gerar(): Promise<string> {
  const cli = carregarYaml<DadosCliente>('exemplo_cliente.yaml');
  const doc = montarDocumento(cli);
  mkdirSync(OUT_DIR, { recursive: true });
  const caminho = join(
    OUT_DIR,
    `Termo_Aceite_UAT_${slug(cli.cliente?.nome)}.docx`,
  );
  writeFileSync(caminho, await Packer.toBuffer(doc));
  return caminho;
}

if (require.main === module) {
  gerar()
    .then((caminho) => console.log(`OK: ${caminho}`))
    .catch((e) => {
      console.error(e);
      process.exitCode = 1;
    });
}
